swarm/pss/snapshot_test.go

```diff
@@ -0,0 +1,433 @@
+			if po > prev {
+				prev = po
+				targets = nil
+				targets = append(targets, nod.ID())
+			} else if po == prev {
+				targets = append(targets, nod.ID())
+			}
+
+			if po >= depth {
+				maxMessages++
+				allowed[i] = append(allowed[i], nod.ID())
+				allowedMsgs[nod.ID()] = append(allowedMsgs[nod.ID()], uint64(i))
+			}
+
+			// a node with the smallest PO (wrt msg) will be the sender
+			if po < smallestPo {
+				smallestPo = po
+				senders[i] = nod.ID()
+			}
+		}
+
+		msgsToReceive += len(targets)
+		for _, id := range targets {
+			recipients[i] = append(recipients[i], id)
+			expectedMsgs[id] = append(expectedMsgs[id], uint64(i))
+		}
+
+		log.Debug("nn for msg", "targets", len(recipients[i]), "msgidx", i, "msg", common.Bytes2Hex(msgAddr[:8]), "sender", senders[i], "senderpo", smallestPo)
+	}
+	log.Debug("msgs to receive", "count", msgsToReceive)
+}
+
+func TestProxNetwork(t *testing.T) {
+	if (*runNodes > 0 && *runMessages == 0) || (*runMessages > 0 && *runNodes == 0) {
+		t.Fatal("cannot specify only one of flags --nodes and --messages")
+	} else if *runNodes > 0 {
+		t.Run(fmt.Sprintf("%d/%d", *runMessages, *runNodes), testProxNetwork)
+	} else {
+		t.Run("1/4", testProxNetwork)
+	}
+}
+
+// This tests generates a sequenced number of messages with random addresses.
+// It then calculates which nodes in the network have the address of each message
+// within their nearest neighborhood depth, and stores them as recipients.
+// Upon sending the messages, it verifies that the respective message is passed to the message handlers of these recipients.
+// It will fail if a recipient handles a message it should not, or if after propagation not all expected messages are handled (timeout)
+func testProxNetwork(t *testing.T) {
+	resetTestVariables()
+	msgCount, nodeCount := getCmdParams(t)
+	handlerContextFuncs := make(map[Topic]handlerContextFunc)
+	handlerContextFuncs[topic] = nodeMsgHandler
+	services := newProxServices(true, handlerContextFuncs, kademlias)
+	snap := readSnapshot(t, nodeCount)
+	sim = simulation.New(services)
+	defer sim.Close()
+	err := sim.Net.Load(&snap)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+	defer cancel()
+	err = sim.WaitTillSnapshotRecreated(ctx, snap)
+	if err != nil {
+		t.Fatalf("failed to recreate snapshot: %s", err)
+	}
+	assingTestVariables(sim, msgCount)
+	result := sim.Run(ctx, runFunc)
+	if result.Error != nil {
+		// context deadline exceeded
+		// however, it might just mean that not all possible messages are received
+		// now we must check if all required messages are received
+		log.Debug("--------------------------------------------------------------------------------", "rcv", msgCnt)
+		if msgCnt < msgsToReceive {
+			t.Fatal(result.Error)
+		}
+	}
+	t.Logf("completed %d", result.Duration)
+}
+
+func sendAllMsgs(sim *simulation.Simulation, msgs [][]byte, senders map[int]enode.ID) {
```

Comment:
why use rpc here? just 'bucket' the pss object

Reply:
actually, i am not sure. i inherited this code, and it works :)

Reply:
@zelig all the send and network tests in `pss_test.go` use RPC. If we are to change this, please let's change it in one go when we clean up the test files, which is a pending task anyway.

However, let's also keep in mind that _most_ usage of `Pss` is indeed through RPC. Thus I don't think it's unreasonable to use RPC calls in these higher level tests.